lib/charms/zookeeper/v0/client.py

```diff
@@ -0,0 +1,221 @@
+import logging
+import re
+import time
+from typing import Any, Dict, Iterable, List, Set, Tuple
+from kazoo.client import KazooClient
+
+logger = logging.getLogger(__name__)
+
+# Kazoo logs are unbearably chatty
+logging.getLogger("kazoo.client").disabled = True
+
+
+class MembersSyncingError(Exception):
+    """Generic exception for when quorum members are syncing data."""
+
+    pass
+
+
+class MemberNotReadyError(Exception):
+    """Generic exception for when a zk unit can't be connected to or is not broadcasting."""
+
+    pass
+
+
+class QuorumLeaderNotFoundError(Exception):
+    """Generic exception for when there are no zk leaders in the app."""
+
+    pass
+
+
+class ZooKeeperManager:
+    """Handler for performing ZK commands."""
+
+    def __init__(self, hosts: List[str], client_port: int = 2181, tries=2, retry_delay=3.0):
+        self.hosts = hosts
+        self.client_port = client_port
+        self.leader = ""
+        self.tries = tries
+        self.retry_delay = retry_delay
+
+        # iterate through all hosts to find current leader
+        for attempt in range(0, self.tries):
+            for host in self.hosts:
+                with ZooKeeperClient(host=host, client_port=client_port) as zk:
+                    response = zk.srvr
+                    if response.get("Mode") == "leader":
+                        self.leader = host
```

Comment:
Only 1 leader thankfully. In theory we could get a split quorum (units A,B* and C*,D,E), but I think I managed to avoid that scenario with code elsewhere.